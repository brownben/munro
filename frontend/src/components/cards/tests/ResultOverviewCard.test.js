import { mount, shallowMount } from '@vue/test-utils'
import ResultsCard from '@/components/cards/ResultOverviewCard'

import axios from 'axios'

jest.mock('axios')

const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockRejectedValue()
})

test('Is a Vue Instance', () => {
  const wrapper = mount(ResultsCard, {
    mocks: {
      $auth: { user: false },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Not Logged In', () => {
  const wrapper = mount(ResultsCard, {
    propsData: {
      result: {
        id: 7,
        event: 'TestEvent',
        points: 43,
        time: 60612,
        position: 5,
        incomplete: false,
        type: '',
      },
    },
    mocks: {
      $auth: { user: false },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Logged In', () => {
  const wrapper = mount(ResultsCard, {
    propsData: {
      result: {
        id: 7,
        event: 'TestEvent',
        points: 49,
        time: 60412,
        position: 2,
        incomplete: true,
        type: 'manual',
      },
    },
    mocks: {
      $auth: { user: true },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Elapsed Time', () => {
  const wrapper = mount(ResultsCard, {
    propsData: {
      result: {
        id: 7,
        event: 'TestEvent',
        points: 49,
        time: 60412,
        position: 2,
        incomplete: true,
        type: 'manual',
      },
    },
    mocks: {
      $auth: { user: false },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.vm.elapsedTime(0)).toBe('')
  expect(wrapper.vm.elapsedTime(1)).toBe('00:01')
  expect(wrapper.vm.elapsedTime(45)).toBe('00:45')
  expect(wrapper.vm.elapsedTime(60)).toBe('01:00')
  expect(wrapper.vm.elapsedTime(83)).toBe('01:23')
  expect(wrapper.vm.elapsedTime(1023)).toBe('17:03')
  expect(wrapper.vm.elapsedTime(3600)).toBe('60:00')
  expect(wrapper.vm.elapsedTime(5117)).toBe('85:17')
})

test('Position Superscript', () => {
  const wrapper = mount(ResultsCard, {
    propsData: {
      result: {
        id: 7,
        event: 'TestEvent',
        points: 49,
        time: 60412,
        position: 2,
        incomplete: true,
        type: 'manual',
      },
    },
    mocks: {
      $auth: { user: false },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.vm.positionSuperscript(0)).toBe('th')
  expect(wrapper.vm.positionSuperscript(1)).toBe('st')
  expect(wrapper.vm.positionSuperscript(2)).toBe('nd')
  expect(wrapper.vm.positionSuperscript(3)).toBe('rd')
  expect(wrapper.vm.positionSuperscript(4)).toBe('th')
  expect(wrapper.vm.positionSuperscript(5)).toBe('th')
  expect(wrapper.vm.positionSuperscript(10)).toBe('th')
  expect(wrapper.vm.positionSuperscript(11)).toBe('th')
  expect(wrapper.vm.positionSuperscript(12)).toBe('th')
  expect(wrapper.vm.positionSuperscript(13)).toBe('th')
  expect(wrapper.vm.positionSuperscript(14)).toBe('th')
  expect(wrapper.vm.positionSuperscript(15)).toBe('th')
  expect(wrapper.vm.positionSuperscript(21)).toBe('st')
  expect(wrapper.vm.positionSuperscript(22)).toBe('nd')
  expect(wrapper.vm.positionSuperscript(23)).toBe('rd')
  expect(wrapper.vm.positionSuperscript(101)).toBe('st')
  expect(wrapper.vm.positionSuperscript(111)).toBe('th')
  expect(wrapper.vm.positionSuperscript(112)).toBe('th')
  expect(wrapper.vm.positionSuperscript(121)).toBe('st')
  expect(wrapper.vm.positionSuperscript(122)).toBe('nd')
})

test('Hide Result', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $auth: { user: true },
    },
    propsData: {
      result: {
        id: 7,
        event: 'EVENT',
        type: '',
      },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.put.mockResolvedValue()
  await wrapper.vm.hideResult()
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith('/api/results/7', {
    rowid: 7,
    action: 'hide',
    event: 'EVENT',
    type: 'hidden',
  })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Result Updated')
})

test('Hide Result - Restore', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $auth: { user: true },
    },
    propsData: {
      result: {
        id: 7,
        event: 'EVENT',
        type: 'hidden',
      },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.put.mockResolvedValue()
  await wrapper.vm.hideResult()
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith('/api/results/7', {
    rowid: 7,
    action: 'hide',
    event: 'EVENT',
    type: null,
  })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Result Updated')
})

test('Hide Result - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $auth: { user: true },
    },
    propsData: {
      result: {
        id: 7,
        event: 'EVENT',
        type: '',
      },
    },
    stubs: ['router-link'],
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.put.mockRejectedValue()
  await wrapper.vm.hideResult()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Hiding Result - Please Try Again'
  )
})

test('Incomplete Result', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $auth: { user: true },
    },
    propsData: {
      result: {
        id: 7,
        event: 'EVENT',
        type: '',
        incomplete: false,
      },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.put.mockResolvedValue()
  await wrapper.vm.incompleteResult()
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith('/api/results/7', {
    rowid: 7,
    action: 'incomplete',
    event: 'EVENT',
    incomplete: true,
  })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Result Updated')
})

test('Incomplete Result - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $auth: { user: true },
    },
    propsData: {
      result: {
        id: 7,
        event: 'EVENT',
        type: '',
        incomplete: false,
      },
    },
    stubs: ['router-link'],
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.put.mockRejectedValue()
  await wrapper.vm.incompleteResult()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Updating Result - Please Try Again'
  )
})
