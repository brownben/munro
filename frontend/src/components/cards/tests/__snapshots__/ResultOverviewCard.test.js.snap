// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Logged In 1`] = `
<section
  class="flex flex-col col-span-2 pt-6 text-center bg-white shadow-md lg:col-span-1 rounded-shape-xl"
>
  <h2
    class="px-4 mb-2 font-bold leading-8 text-25xl md:text-3xl font-heading text-main-700"
  >
    
      
    
  </h2>
   
  <div
    class="flex flex-col flex-wrap justify-around w-full sm:flex-row"
  >
    <div
      class="inline-block mx-4 mt-6 mb-2 leading-8 whitespace-no-wrap sm:mt-8 font-heading text-main-900 sm:mb-0"
    >
      <p
        class="text-6xl font-bold xl:text-8xl"
      >
        
          1006:52
        
      </p>
       
      <p
        class="inline-block mt-4 text-opacity-50 uppercase text-main-900"
      >
        
        Time
      
      </p>
    </div>
     
    <div
      class="inline-block mx-4 mt-8 leading-8 whitespace-no-wrap sm:mt-8 font-heading text-main-900"
    >
      <p
        class="text-6xl font-bold xl:text-8xl"
      >
        
          *
        
      </p>
       
      <p
        class="inline-block mt-4 text-opacity-50 uppercase sm:mt-4 text-main-900"
      >
        
        Position
      
      </p>
    </div>
  </div>
   
  <div>
    <p
      class="mt-2 text-xl text-opacity-75 sm:mt-0 text-main-900 font-heading"
    >
      <span
        class="text-2xl font-bold"
      >
         49
      </span>
       Points
    
    </p>
     
    <router-link-stub
      class="mx-6 mt-4 mb-4 sm:mt-3 button"
      to="/events/TestEvent/results"
    >
      
      View Results
    
    </router-link-stub>
  </div>
   
  <div
    class="flex-grow w-full px-4 pt-4 pb-4 mt-2 shadow md:px-6 bg-main-100 rounded-shape-xl"
  >
    <h3
      class="mb-1 text-2xl font-bold text-center text-main-800 font-heading"
    >
      
      Admin Actions
    
    </h3>
     
    <div
      class="flex flex-row flex-wrap justify-around px-6"
    >
      <p
        class="mx-2 mb-1"
      >
        <b
          class="mr-2 text-main-800"
        >
          Id:
        </b>
        
        7
      
      </p>
       
      <p
        class="mx-2 mb-1"
      >
        <b
          class="mr-2 text-main-800"
        >
          Type:
        </b>
        
        manual
      
      </p>
       
      <p
        class="mx-2 font-bold text-main-800 font-heading"
      >
        
        Incomplete Result
      
      </p>
    </div>
     
    <div
      class="w-full mt-2"
    >
      <button
        class="button-dark"
      >
        Mark as Complete
      </button>
       
      <button
        class="button-dark"
      >
        Hide Result
      </button>
    </div>
  </div>
</section>
`;

exports[`Renders Correctly - Not Logged In 1`] = `
<section
  class="flex flex-col col-span-2 pt-6 text-center bg-white shadow-md lg:col-span-1 rounded-shape-xl"
>
  <h2
    class="px-4 mb-2 font-bold leading-8 text-25xl md:text-3xl font-heading text-main-700"
  >
    
      
    
  </h2>
   
  <div
    class="flex flex-col flex-wrap justify-around w-full sm:flex-row"
  >
    <div
      class="inline-block mx-4 mt-6 mb-2 leading-8 whitespace-no-wrap sm:mt-8 font-heading text-main-900 sm:mb-0"
    >
      <p
        class="text-6xl font-bold xl:text-8xl"
      >
        
          1010:12
        
      </p>
       
      <p
        class="inline-block mt-4 text-opacity-50 uppercase text-main-900"
      >
        
        Time
      
      </p>
    </div>
     
    <div
      class="inline-block mx-4 mt-8 leading-8 whitespace-no-wrap sm:mt-8 font-heading text-main-900"
    >
      <p
        class="text-6xl font-bold xl:text-8xl"
      >
        
          5
          
        <sup
          class="relative inline-block -ml-3 text-3xl font-normal text-opacity-75 -top-7 text-main-900"
        >
          
            th
          
        </sup>
      </p>
       
      <p
        class="inline-block mt-4 text-opacity-50 uppercase sm:mt-4 text-main-900"
      >
        
        Position
      
      </p>
    </div>
  </div>
   
  <div>
    <p
      class="mt-2 text-xl text-opacity-75 sm:mt-0 text-main-900 font-heading"
    >
      <span
        class="text-2xl font-bold"
      >
         43
      </span>
       Points
    
    </p>
     
    <router-link-stub
      class="mx-6 mt-4 mb-4 sm:mt-3 button"
      to="/events/TestEvent/results"
    >
      
      View Results
    
    </router-link-stub>
  </div>
   
  <!---->
</section>
`;
